core/cap-0051.md

```diff
@@ -260,18 +260,22 @@ func $serialize_to_binary(param $val i64) (result i64)
 /// Traps if the deserialization fails for any reason.
 func $deserialize_from_binary(param $obj_bin i64, param $type u64) (result i64)
 
-/// Given a host binary object $obj_bin, copy a segment of its memory specified at $offset with length $len into guest linear memory at $pos. 
+/// Given a host binary object $obj_bin, copies a segment of its memory specified at $offset with length $len into the guest linear memory at $pos. 
 /// Traps if either the binary object or the guest linear memory doesn't have enough bytes.
-func $binary_copy_to_guest_mem(param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64)
+func $binary_memcpy_to_guest(param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64)
 
-/// Copy a segment of the guest's linear memory specified at position $pos with length $len, into a host binary object $obj_bin at $offset. 
+/// Copies a segment of the guest linear memory specified at position $pos with length $len, into a host binary object $obj_bin at $offset. The host binary binary may grow in size to accomadate the new bytes. 
 /// Returns handle to the new binary array.
-/// Traps if either the binary object or the guest linear memory doesn't have enough bytes.
-func $binary_copy_from_guest_mem(param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64) (result i64)
+/// Traps if the guest linear memory doesn't have enough bytes.
+func $binary_memcpy_from_guest(param $obj_bin i64) (param $offset u64) (param $pos u64) (param $len u64) (result i64)
```

Comment:
❓ There doesn't seem to be anything functional to distinguish `binary_memcpy_from_guest` from the new `binary_new_from_memcpy`. Both functions return a new binary object because all host objects are immutable, so the passed in `obj_bin` in the first function won't get used. Is that correct? If so, we can omit `binary_memcpy_from_guest` in favor of `binary_new_from_memcpy`.

Reply:
Also, regarding which one to keep if we could only have one, I thought `binary_copy_from_linear_memory` is more general than `binary_new_from_linear_memory` as it can be replicated by `binary_new` + `binary_copy_from_linear_memory`. 
But I also see `binary_copy_from_linear_memory` can be replicated by `binary_copy_to_linear_memory` + `binary_new_from_linear_memory` so I guess this is subjective. But since we already have a `_copy_to_`, I figured it makes more sense to keep `_copy_from_`.

Reply:
Ah yes I see what you mean, the existing function actually has more functionality because it allows you to combine an existing binary with some linear memory. I think it's fine to keep both, I misunderstood when I read it. 

Also, does this mean we need versions for `insert`, `push` and `pop` too, where linear memory is a source or destination?

Reply:
Quite possibly, whichever set needed to replicated the existing binary functionality with ranged capability. I will think about those and propose them next. The ones @graydon suggested so far are `memcpy`, `memset` and `memchar` (-equivalent).

Reply:
> I think it's fine to keep both

Cool, I added the `_new_` back, just because I already have the impl for it in the contract-env PR.